src/op_analytics/cli/subcommands/pulls/agora/delegates.py

```diff
@@ -0,0 +1,207 @@
+                lower_bound.add(lo)
+                upper_bound.add(hi)
+                if diff < 1000:
+                    self.min_offset = lo
+                    self.max_offset = hi
+                    return hi
+                else:
+                    hi = midpoint
+            else:
+                log.warning("Unexpected state encountered in binary search.")
+                hi = midpoint
+
+    def concurrent_fetch(self):
+        self.max_offset = self.binary_search_max_offset()
+        offsets = list(range(0, self.max_offset, self.limit))[:10]
+        data = run_concurrently(
+            function=lambda offset: self.request(offset).data,
+            targets=offsets,
+            max_workers=self.max_workers,
+        )
+        return list(itertools.chain.from_iterable(data.values()))
+
+
+def pull_delegates():
+    """Pull and write agora delegates data."""
+    p = Paginator(url=f"{BASE_URL}/delegates", params={"sort": "voting_power"}, max_workers=32)
+    delegate_data = p.concurrent_fetch()
+
+    normalized_data = []
+    for item in delegate_data:
+        address = item.get("address", "")
+        voting_power = item.get("votingPower", {})
+        citizen = item.get("citizen", False)
+        statement = item.get("statement", {})
+        payload = statement.get("payload", {})
+
+        # Raise an exception if 'voting_power' is not a dict
+        if not isinstance(voting_power, dict):
+            raise ValueError(
+                f"Expected 'votingPower' to be a dict for address {address}, got {type(voting_power)}"
+            )
+
+        flattened = {
+            "address": address,
+            "voting_power_total": voting_power["total"],
+            "voting_power_direct": voting_power["direct"],
+            "voting_power_advanced": voting_power["advanced"],
+            "is_citizen": citizen,
+            "statement_signature": statement.get("signature", ""),
+            "statement_created_at": statement.get("created_at", ""),
+            "statement_updated_at": statement.get("updated_at", ""),
+            "statement_text": payload.get("delegateStatement", ""),
+            "twitter": payload.get("twitter", ""),
+            "discord": payload.get("discord", ""),
+        }
+
+        normalized_data.append(flattened)
+
+    # Construct polars DataFrame directly
+    df = pl.DataFrame(normalized_data)
+
+    # Deduplicate by address if needed
+    initial_count = df.height
+    df = df.unique(subset=["address"])
+    final_count = df.height
+    if final_count < initial_count:
+        log.warning(f"Dropped {initial_count - final_count} duplicate rows based on 'address'.")
+
+    # Add a dt column for partitioning
+    df = df.with_columns(pl.lit(now_dt()).alias("dt"))
+
+    # Write to GCS using Agora's write
+    Agora.DELEGATES.write(dataframe=df, sort_by=["voting_power_total"])
+
+    return AgoraDelegates(delegates_df=df)
+
+
+# Todo: move to app.py
```

Comment:
Let's go ahead and move this to app.py. I was going to test it locally and realized I could not yet access the code from the CLI.